File: 78-subsets/78-subsets.java
class Solution {
    List<List<Integer>> subs = new ArrayList<>();
    int k; //size of current combination
    int n;
    
    
    public void backtrack(int start, ArrayList<Integer> temp, int[] nums) {
        if (temp.size() == k) {
            subs.add(new ArrayList(temp));
            return;
        }
        for (int i = start; i < n; i++) {
            temp.add(nums[i]);
            backtrack(i + 1, temp, nums);
            temp.remove(temp.size() - 1);
        }
    }
    
    public List<List<Integer>> subsets(int[] nums) {
        n = nums.length;
        for (k = 0; k < n + 1; k++) {
            backtrack(0, new <Integer>ArrayList(), nums);
        }
        return subs;
        
    }
}

File: 78-subsets/78-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        sol = []
        subset = []
        def backtrack(i):
            if (i >= len(nums)):
                sol.append(subset.copy())
                return
            #left side of tree, add curr number
            subset.append(nums[i])
            backtrack(i + 1)
            
            #right side of tree, add nothing
            subset.pop()
            backtrack(i + 1)
        backtrack(0)
        return sol
        